jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java

```diff
@@ -453,17 +457,21 @@ else if (networkChannel instanceof DatagramChannel)
             boolean connected = true;
             if (channel instanceof SocketChannel socketChannel)
             {
+                final SocketAddress socketAddress = address;
                 boolean blocking = isConnectBlocking() && address instanceof InetSocketAddress;
                 if (LOG.isDebugEnabled())
                     LOG.debug("Connecting {} to {}", blocking ? "blocking" : "non-blocking", address);
                 if (blocking)
                 {
+                    listeners.forEach(listener -> listener.onConnectBegin(socketChannel, socketAddress));
```

Comment:
These calls should be wrapped, since we do not want an exception thrown from `onConnectBegin()` to stop the connection.
Make a method, say `connectBegin()`, use a `for` loop instead of `forEach()`, and wrap `listener.onConnectBegin()` into `try/catch(Throwable)`.
And same for the other events.

Reply:
What is the reason of this solution?
`ConnectListener` methods don't throw an exception.

Reply:
`ConnectListener` implementations are user code, and it can throw any `RuntimeException`, but we do not want `ClientConnector` internal working to be affected by that if it happens.

Reply:
Yeah, it is pretty  much useful!
Thank you for your tips!